lib/compilers/tic2000.js

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2019, Compiler Explorer Authors
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import path from 'path';
+
+import _ from 'underscore';
+
+import {BaseCompiler} from '../base-compiler';
+import {TiC2000AsmParser} from '../parsers/asm-parser-tic2000';
+
+export class TIC2000 extends BaseCompiler {
+    static get key() {
+        return 'tic2000';
+    }
+
+    constructor(info, env) {
+        super(info, env);
+        this.outputFilebase = this.compileFilename.split('.')[0];
+        this.asm = new TiC2000AsmParser(this.compilerProps);
+    }
+
+    optionsForFilter(filters, outputFilename) {
+        const options = ['-g', '-c', '-n', '--output_file=' + outputFilename];
+
+        filters.preProcessLines = _.bind(this.preProcessLines, this);
+
+        return options;
+    }
+
+    getOutputFilename(dirPath, outputFilebase) {
+        return path.join(dirPath, `${outputFilebase}.asm`);
+    }
+
+    exec(compiler, args, options_) {
+        const options = Object.assign({}, options_);
+        options.env = Object.assign({}, options.env);
+
+        if (this.compiler.includePath) {
+            options.env['INCLUDE'] = this.compiler.includePath;
+        }
+        if (this.compiler.libPath) {
+            options.env['LIB'] = this.compiler.libPath;
+        }
+        for (const [env, to] of this.compiler.envVars) {
+            options.env[env] = to;
+        }
+
+        options.env['TMP'] = process.env.TMP;
+
+        return super.exec(compiler, args, options);
+    }
+
+    getExtraAsmHint(asm) {
```

Comment:
It feels like there ought to be a simpler way to do this. But without tests, I can't make any suggestions.

Reply:
I wrote this way back in the latter half of 2019, so I'm sure things have changed. I'll see what I can do to improve it.

Reply:
I believe I pulled this from the pascal implementation, which added the `.file` and `.loc` lines to allow the source to be tied to the assembly.

I looked to see if I could replace the `sourceTag` regex in the parser with my own, but it expects a file number and lookup, which I don't have. I'm given the full name, which is why the extra lines were added.

I did simplify it to a regex match, however. Hopefully that makes it easier to understand, and I'll write a test for that as well. (Perhaps the `pascal-tests.js` file would be a good place to start as an example?)